### PR TITLE
Divide eav attribute cache per website

### PR DESCRIPTION
**Pull request by vshcheglov**
### Description (*)
We have an ability to choose different attribute settings per website in Magento admin, but have no per website eav attribute cache separation. So our attribute settings will be depended on which website will be opened first after eav cache cleaning.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#10978

### Manual testing scenarios (*)
1. Open clean Magento instance

Creating new website and store
1. Go to Stores -> Settings -> All Stores
1. Select Create Website
1. Create a new website
1. Select Create Store
1. Create a new store and assign it to a newly created website
1. Select Create Store View
1. Create a new enabled store view and assign it to newly created store
1. Remember store view code
1. Go to System -> Cache Management
1. Flush Cache Storage

Filling required settings
1. Go to Stores -> Configuration -> Customers -> Customer Configuration -> Name and Address Options
1. Select Main Website scope
1. Set Date of Birth to No
1. Set Show Tax/VAT Number to Optional
1. Save config
1. Select new website scope
1. Set Date of Birth to Optional
1. Set Show Tax/VAT Number to No
1. Save config

Enabling store code in urls for fast website switching
1. Select Default Config scope
1. Go to General -> Web -> Url Options
1. Change Add Store Code to Urls to Yes
1. Save config

Reproducing the bug
1. Go to System -> Cache Management
1. Flush cache storage
1. Important! First of all, go to Main Website Storefront Create Account Page
1. Go to https://yourmagento.com/default/customer/account/create/
1. We see Tax/Vat Number field and don't see Dob field
1. Then go to new created website Create Account Page
1. Go to https://yourmagento.com/new-store-view-code/customer/account/create/

Expected:
1. We don't see Tax/Vat Number field and see Dob field

Actual:
1. We see Tax/Vat Number field and don't see Dob field

Then another case:
1. Flush cache storage
1. Important! First of all, go to New Website Storefront Create Account Page
1. Go to yourmagento.com/new-store-view-code/customer/account/create/
1. We don't see Tax/Vat Number field and see Dob field (as expected sooner above)
1. Now go to Main Website Storefront create account page
1. Go to https://yourmagento.com/default/customer/account/create/

Expected:
1. We see Tax/Vat Number field and don't see Dob field

Actual:
1. We don't see Tax/Vat Number field and see Dob field

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by vshcheglov** [comment]
@magento give me 2.5-develop instance

**Comment by rikwillems** [comment]
I just added a suggestion to the PR.

**Comment by rikwillems** [comment]
@vshcheglov please see my two PRs to yours. The current implementation is not working for me without these changes.

**Comment by vshcheglov** [comment]
> @vshcheglov please see my two PRs to yours. The current implementation is not working for me without these changes.

Yeah, I missed some changes. Thank you!

**Comment by danielrussob** [comment]
@magento give me 2.5-develop instance

**Comment by danielrussob** [comment]
@magento give me test instance

**Comment by aleron75** [comment]
@magento run all tests

**Comment by danielrussob** [comment]
I have reproduced the issue on vanilla magento 2.5 instance

**Comment by aleron75** [comment]
Hello @vshcheglov!

Thanks for your contribution.

Due to [Magento Definition of Done](https://devdocs.magento.com/contributor-guide/contributing_dod.html#automated-tests), all code must be covered by tests.

For this specific case, this fix should be covered by automated tests with the scenario which leads to an issue: tests should fail on the mainline and pass with this fix.

A good candidate would be extending the `\Magento\Eav\Test\Unit\Model\ConfigTest` unit test.

Thank you!

**Comment by danielrussob** [comment]
@magento give me test instance

**Comment by aleron75** [comment]
One last thing @vshcheglov: generally only features and tasks with BIC are allowed to be introduced into the `2.5-develop` branch. 

All the bugfix should be delivered to the `2.4-develop` branch;  all fixes delivered to `2.4-develop` will be automatically ported to `2.5-develop`.

So, please, submit this bugfix to the `2.4-develop` branch.

Thanks.

**Comment by Nazar65** [comment]
Hi @vshcheglov  do we need to do the same with child's products for configurable, it also uses same cache for front and back, and also store here : -> https://github.com/magento/magento2/blob/0a05fa3da4c645926af0e6d8dea1c81d819be8fc/app/code/Magento/ConfigurableProduct/Model/Product/Type/Configurable.php#L1436

**Comment by vshcheglov** [comment]
@aleron75 
> So, please, submit this bugfix to the 2.4-develop branch.

patch-2 branch is based on 2.5-develop branch so we should not merge it to 2.4

I prepared a new branch based on 2.4.develop branch but don't see any ability to change patch-2 branch to a new branch.
How to deal with it? I can create a new pull request with a new branch.

@Nazar65 
> Hi @vshcheglov do we need to do the same with child's products for configurable, it also uses same cache for front and back, and also store here : ->

I am not sure that it is related to the current issue.

File: app/code/Magento/Eav/Model/Config.php [hunk 1]
```diff
@@ -960,4 +970,14 @@ private function initAttributesFromCache(Type $entityType)
         }
         return false;
     }
+
+    /**
+     * Get current website id
+     * 
+     * @return int
+     */
+    private function getWebsiteId()
```
Comment (sidolov) [inline app/code/Magento/Eav/Model/Config.php]: Please, add return type

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by vshcheglov** [comment]
Added a pull request with additional changes based on 2.4-develop branch being merged to it.
https://github.com/magento/magento2/pull/34856

**Pull request closed without merging**